vdj_ann/src/transcript.rs

```diff
@@ -7,187 +7,252 @@ use crate::refx::RefData;
+    contig_status.in_frame = Some(inframe_pair.is_some());
+
+    contig_status.no_premature_stop = if let Some((vstart, jstop)) = inframe_pair {
+        Some(
+            !(vstart.tig_start..jstop.tig_stop - 3)
+                .step_by(3)
+                .any(|j| have_stop(contig, j)),
+        )
+    } else {
+        None
+    };
+
+    let mut cdr3 = Vec::<(usize, Vec<u8>, usize, usize)>::new();
+    get_cdr3_using_ann(contig, reference, ann, &mut cdr3);
+    contig_status.has_cdr3 = Some(!cdr3.is_empty());
+
+    contig_status.has_expected_size =
+        if let (Some((vstart, jstop)), Some(cdr3)) = (inframe_pair, cdr3.first()) {
+            let expected_len = (refs[vstart.ref_id].len() + refs[jstop.ref_id].len()) as i32
+                + (3 * cdr3.1.len() as i32)
+                - 20;
+            let observed_len = jstop.tig_stop as i32 - vstart.tig_start as i32;
+            let delta = expected_len - observed_len;
+            let min_delta = if vdj_chain == VdjChain::IGH {
+                MIN_DELTA_IGH
+            } else {
+                MIN_DELTA
+            };
             if delta < min_delta || delta > MAX_DELTA {
-                too_large = true;
-                if logme {
-                    fwriteln!(log, "delta too large");
-                }
+                Some(false)
+            } else {
+                Some(true)
             }
-        }
-        let mut misordered = false;
-        for j1 in 0..ann.len() {
-            let t1 = ann[j1].2 as usize;
-            for j2 in j1 + 1..ann.len() {
-                let t2 = ann[j2].2 as usize;
-                if (refdata.is_j(t1) && refdata.is_v(t2))
-                    || (refdata.is_j(t1) && refdata.is_u(t2))
-                    || (refdata.is_j(t1) && refdata.is_d(t2))
-                    || (refdata.is_v(t1) && refdata.is_u(t2))
-                    || (refdata.is_c(t1) && !refdata.is_c(t2))
-                {
-                    misordered = true;
-                }
-            }
-        }
-        if misordered && logme {
-            fwriteln!(log, "misordered");
-        }
-        if !full && logme {
-            fwriteln!(log, "not full");
-        }
-        if full && !too_large && !misordered {
-            return true;
-        }
+        } else {
+            None
+        };
+
+    let observed_order: Vec<i32> = annotation
+        .into_iter()
+        .map(|a| reference.segtype[a.ref_id])
+        .map(|s| match s {
+            "U" => 0,
+            "V" => 1,
+            "D" => 2,
+            "J" => 3,
+            "C" => 4,
+            _ => panic!("Invalid segtype"),
+        })
+        .collect();
+    let mut expected_order = observed_order.clone();
+    expected_order.sort();
+    contig_status.correct_ann_order = Some(observed_order == expected_order);
+
+    Some(contig_status)
+}
+
+pub fn is_productive_contig(
+    b: &DnaString,
+    refdata: &RefData,
+    ann: &[(i32, i32, i32, i32, i32)],
```

Comment:
Not necessary in this PR, but we should at some point in the future replace these annotation tuples with structs.